### PR TITLE
feat: add better error reporting for COM/IOM connections

### PR DESCRIPTION
**Pull request by scottdover**
**Summary**
This adds better error support for IOM & COM connections. Specifically, it provides _better_ errors for:
 - being on the wrong plaform (i.e. Mac)
 - using an incorrect username / password
 - using an incorrect host
 - using an incorrect port (or having a connection where a sas server isn't present)
 - trying to run code that will trigger transcoding errors

**Testing**
 - [x] test being on the wrong plaform (i.e. Mac)
 - [x] test using an incorrect username / password
 - [x] test using an incorrect host
 - [x] test using an incorrect port (or having a connection where a sas server isn't present)
 - [x] test trying to run code that will trigger transcoding errors

### CODE REVIEW

File: client/src/connection/itc/const.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Copyright © 2024, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { l10n } from "vscode";
+
+export const humanReadableMessages = [
```
Comment (scottdover) [inline client/src/connection/itc/const.ts:5]: Hey @clangsmith . Do we think it's worth having a resource to expand on what these error messages mean? Or perhaps we can tweak these messages to elaborate? 

Really, the only messages that look like they _could_ require more explanation are the errors on line 18 and 28 (imo).

File: client/src/connection/itc/index.ts [hunk 2]
```diff
@@ -13,10 +20,13 @@ import {
...
 import { extractOutputHtmlFileName } from "../../components/utils/sasCode";
 import { Session } from "../session";
+import { LineParser } from "./LineParser";
+import { ERROR_END_TAG, ERROR_START_TAG } from "./const";
 import { scriptContent } from "./script";
 import { LineCodes } from "./types";
+import { decodeEntities } from "./util";
 
-const SECRET_STORAGE_NAMESPACE = "ITC_SECRET_STORAGE";
+const SECRET_STORAGE_NAMESPACE = "ITC_SECRET_STORAGE-1";
```
Comment (scnwwu) [inline client/src/connection/itc/index.ts:29]: is this change necessary?

File: client/src/connection/itc/index.ts [hunk 3]
```diff
@@ -280,22 +300,80 @@ export class ITCSession extends Session {
...
-        ),
-      ),
+      new Error(this.fetchHumanReadableErrorMessage(errorMessage)),
     );
+
     // If we encountered an error in setup, we need to go through everything again
-    if (/Setup error/.test(msg)) {
+    const fatalErrors = [/Setup error/, /powershell\.exe: command not found/];
+    if (fatalErrors.find((regex) => regex.test(errorMessage))) {
+      // If we can't even run the shell script (i.e. powershell.exe not found),
```
Comment (scnwwu) [inline client/src/connection/itc/index.ts:316]: if the system language is not English, it may get translated "not found", isn't it? not sure if there's any better way to check this

File: client/src/connection/itc/LineParser.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+export class LineParser {
+  protected errorLines: string[] = [];
+  protected capturingError: boolean = false;
+  protected startTag: string;
+  protected endTag: string;
+
+  public constructor(startTag: string, endTag: string) {
+    this.startTag = startTag;
+    this.endTag = endTag;
+  }
```
Comment (scnwwu) [inline client/src/connection/itc/LineParser.ts:13]: nit: just a note that this can be simplified to
`public constructor(protected startTag: string, protected endTag: string) {}`
https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties
But I'm fine if you prefer this

File: client/src/connection/itc/index.ts [hunk 5]
```diff
@@ -280,22 +300,80 @@ export class ITCSession extends Session {
...
       this._workDirectory = undefined;
     }
   };
 
+  private fetchHumanReadableErrorMessage = (msg: string): string => {
+    const atLineIndex = msg.indexOf("At line");
+    const errorMessage = atLineIndex ? msg.slice(0, atLineIndex) : msg;
+
+    // Dump error to console
+    console.warn("shellProcess stderr: " + errorMessage);
```
[the remark this answers is not in the document]
Comment (scottdover) [reply client/src/connection/itc/index.ts:333]: Hmm. Perhaps. Fwiw, this was just copied from another location. That said, I think there was _some_ duplication in the console. Let me clean that up and see if we still see all the necessary information

File: client/src/connection/itc/index.ts (hunk 2 above)
Comment (scottdover) [reply client/src/connection/itc/index.ts]: d'oh. Updated in https://github.com/sassoftware/vscode-sas-extension/pull/842/commits/d7f28f15d5c4ec6925932a8f3ac546b28e816ecd

File: client/src/connection/itc/LineParser.ts (hunk 4 above)
Comment (scottdover) [reply client/src/connection/itc/LineParser.ts]: I updated it to this format in https://github.com/sassoftware/vscode-sas-extension/pull/842/commits/d7f28f15d5c4ec6925932a8f3ac546b28e816ecd . That said, I'm not entirely sure how I feel about it. 

I think my reservation with this approach is that it requires removing the class attributes from the top of the class. For me, this makes it a little less clear what attributes are defined. That said, it's super minor and as long as the constructor is close enough to the rest of the attributes, it seems fine

File: client/src/connection/itc/index.ts (hunk 3 above)
Comment (scottdover) [reply client/src/connection/itc/index.ts:316]: That's a good point. We _could_ just isolate this to check `powershell\.exe`. I wonder if that'd be problematic. Fwiw, the _only_ error I've seen with `powershell\.exe` as part of the message is this one 🤷

Comment (scottdover) [reply client/src/connection/itc/index.ts:316]: Okay, I simplified this in https://github.com/sassoftware/vscode-sas-extension/pull/842/commits/58babb4f1a20be6cc3590605cf35b0bfa2a5bbe4. It _seems_ to work well for running code successfully and w/ errors. 

That said, if we think it's _too_ generic, I can roll it back and just let this error fall through to the "default" error

File: client/src/connection/itc/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (scottdover) [reply client/src/connection/itc/index.ts:333]: So, there is some duplication here. However, this line prints the raw error from powershell (which could be useful for debugging), while the other only prints the formatted error. I think there's value in leaving this here.

**Review by scnwwu — APPROVED** [review]

**Comment by ShermanWang0324** [comment]
testing_results

**Comment by ShermanWang0324** [comment]
I have tested all the scenarios listed, all fixed messages are captured as above screenshot.

**Review by ShermanWang0324 — APPROVED** [review]
